File: 160 Intersection of Two Linked Lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) {
 *         val = x;
 *         next = null;
 *     }
 * }
 */
public class Solution {
    public ListNode getIntersectionNode(ListNode headA, ListNode headB) {
        if(headA == null || headB == null) return null;
        int lenA = 0, lenB = 0;
        ListNode iteA = headA, iteB = headB;
        // Calculate length of linked list A and linked list B
        while (iteA != null) {
            iteA = iteA.next;
            lenA++;
        }
        while (iteB != null) {
            iteB = iteB.next;
            lenB++;
        }
        // Let the longer linked list go through the diff nodes first
        int diff = lenA - lenB;
        iteA = headA;
        iteB = headB;
        if (diff > 0) {
            while (diff > 0) {
                iteA = iteA.next;
                diff--;
            }
        } else {
            while (diff < 0) {
                iteB = iteB.next;
                diff++;
            }
        }
        // Go one by one and compare, then we'll get the intersected node. (Or meet null at the end)
        while(iteA != null && iteB != null) {
            if (iteA != iteB) {
                iteA = iteA.next;
                iteB = iteB.next;
            } else {
                return iteA;
            }
        }
        return null;
    }
}

File: 160 Intersection of Two Linked Lists.py
# Definition for singly-linked list.
# class ListNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        iteA, iteB = headA, headB
        lenA, lenB = 0, 0
        while iteA is not None:
            lenA += 1
            iteA = iteA.next
        while iteB is not None:
            lenB += 1
            iteB = iteB.next
        diff = lenA - lenB
        iteA, iteB = headA, headB
        if diff > 0:
            for _ in range(diff):
                iteA = iteA.next
        else:
            for _ in range(-diff):
                iteB = iteB.next
        while iteA and iteB:
            if iteA == iteB:
                return iteA
            iteA, iteB = iteA.next, iteB.next
        return None

''' Solution from icrtiou
class Solution(object):
    def getIntersectionNode(self, headA, headB):
        """
        :type head1, head1: ListNode
        :rtype: ListNode
        """
        if not headA or not headB: return None
        iteA, iteB = headA, headB
        
        while iteA != iteB:
            iteA = headB if not iteA else iteA.next
            iteB = headA if not iteB else iteB.next
        
        return iteA
'''